autoload/crystalline.vim

```diff
@@ -382,7 +439,13 @@ function! crystalline#sep(group_a, group_b, ch, left) abort
   if a:left == 0 && a:group_a ==# 'TabType' && index(get(g:, 'crystalline_tab_type_fake_separators', []), a:group_b) >= 0
     let l:sep_item = g:crystalline_tab_separator
   else
-    let l:sep_item = '%#Crystalline' . crystalline#get_sep_group(a:group_a, a:group_b) . '#' . a:ch
+    let l:sep_group = crystalline#get_sep_group(a:group_a, a:group_b)
+    " Create if it doesn't exist
+    if !hlexists(l:sep_group)
```

Comment:
Sorry I'm a little late catching this, but this won't work for changing the theme on the fly via `set_theme()`, will it? Once the highlight group has been set, won't `hlexists` always return true? Unless if I'm missing where groups get cleared.

This is a problem for switching colorschemes, or even switching brightness on the fly, which gruvbox and maybe other colorschemes support.

Reply:
> This is a problem for switching colorschemes, or even switching brightness on the fly, which gruvbox and maybe other colorschemes support.

True, but I'm afraid I have no experience with themes that do fancy stuff like that. Do you know how gruvbox implements its brightness/contrast changes? Am I correct in thinking that these are basically separate colorschemes? 

Anyway, from what I can tell, all separator HGs match the regex `Crystalline.\{-}To.\{-}` mentioned in https://github.com/rbong/vim-crystalline/pull/21#discussion_r396452934. It appears to me we could just recompute all separator HGs in a `ColorScheme` autocommand if brightness/contrast changes are just implemented as separate colorschemes.

Reply:
I skimmed over the [Gruvbox source](https://github.com/morhetz/gruvbox/blob/master/colors/gruvbox.vim) and from what I can tell the theme just sets its HGs once per colorscheme change, which means recomputing separator HGs on a `Colorscheme` autocmd should suffice.

Reply:
Yes, updating on `ColorScheme` should do it. In the latest commit on master, you can do it in `apply_current_theme`, which is run on both `ColorScheme` and `set_theme`.

Reply:
I can think of two ways to get the names and attributes of all separator HGs:

1. Save the groups of the separator HG into a variable upon creation (i.e. in `crystallline#sep()`).

   This requires that the user *only* uses separators via the interface provided by Crystalline, otherwise they won't be recomputed on a colorscheme change.

2. Get the list of separator highlight groups by parsing the output of `:highlight` or `getcompletion('Crystalline', 'highlight')`.

   This is butt-ugly: parsing `:highlight` feels like working around the lack of a proper interface, and parsing `getcompletion()` feels like abuse of the completion system. It also has a potential performance impact compared to 1., though I'm not sure how relevant that is, considering we're a) only recomputing a few HGs, and b) only on a colorscheme change.

Afterwards, we just run `generate_sep_hi(group_a, group_b)` on the separator HGs.

Reply:
>This requires that the user only uses separators via the interface provided by Crystalline, otherwise they won't be recomputed on a colorscheme change.

I *think* it should be fine as long as you register groups inside of `generate_sep_hi` and remove them in `apply_current_theme`, if plugin creators or users don't use the function they will need to clean up their highlight groups in `CrystallineSetTheme` anyways.